util/compression.h

```diff
@@ -133,12 +133,129 @@ class ZSTDUncompressCachedData {
 
 namespace rocksdb {
 
-// Instantiate this class and pass it to the uncompression API below
+// Holds dictionary and related data, like ZSTD's digested dictionary.
+// Can be used without calling `Init()`, in which case it'll behave as if it
+// were initialized with an empty dictionary.
+struct CompressionDict {
+  enum class Mode {
+    kUninit,
+    kCompression,
+    kUncompression,
+  };
+#if ZSTD_VERSION_NUMBER >= 700
+  union {
+    ZSTD_CDict* zstd_cdict_;
+    ZSTD_DDict* zstd_ddict_;
+  };
+#endif  // ZSTD_VERSION_NUMBER >= 700
+  Mode mode_ = Mode::kUninit;
+  Slice dict_;
+
+ public:
+  static const CompressionDict& GetEmptyDict() {
+    static const CompressionDict empty_dict{};
+    return empty_dict;
+  }
+
+#if ZSTD_VERSION_NUMBER >= 700
+  void Init(Slice dict, Mode mode, CompressionType type, int level = -1) {
+#else   // ZSTD_VERSION_NUMBER >= 700
+  void Init(Slice dict, Mode mode, CompressionType /*type*/,
```

Comment:
I made this `Init` function instead of doing initialization in the constructor because in a future PR we will need to initialize `CompressionDict` in the middle of `BlockBasedTableBuilder`'s lifetime (and `BlockBasedTableBuilder` is the `CompressionDict` owner). In retrospect we could've alternatively made `BlockBasedTableBuilder` hold a `std::unique_ptr<CompressionDict>` and then construct the `CompressionDict` in the middle of `BlockBasedTableBuilder`'s lifetime.

Reply:
LMK if you have an opinion on which way this should be. I slightly lean towards the latter approach, but am reluctant to rework the code...

Reply:
Using Init() seems fine to me.